File: GLLara/GLLSceneDrawer.m
//
//  GLLSceneDrawer.m
//  GLLara
//
//  Created by Torsten Kammer on 03.09.12.
//  Copyright (c) 2012 Torsten Kammer. All rights reserved.
//

#import "GLLSceneDrawer.h"

#import <OpenGL/gl3.h>
#import <OpenGL/gl3ext.h>

#import "NSColor+Color32Bit.h"
#import "GLLAmbientLight.h"
#import "GLLCamera.h"
#import "GLLDirectionalLight.h"
#import "GLLDocument.h"
#import "GLLDrawState.h"
#import "GLLItem.h"
#import "GLLNotifications.h"
#import "GLLRenderParameter.h"
#import "GLLResourceManager.h"
#import "GLLView.h"
#import "simd_matrix.h"
#import "simd_project.h"
#import "GLLTiming.h"

#import "GLLara-Swift.h"

@interface GLLSceneDrawer ()
{
	NSMutableArray *itemDrawers;
	id managedObjectContextObserver;
    id drawStateNotificationObserver;
	GLLSkeletonDrawer *skeletonDrawer;
}

- (void)_addDrawerForItem:(GLLItem *)item;

@end

@implementation GLLSceneDrawer

- (id)initWithDocument:(GLLDocument *)document;
{
	if (!(self = [super init])) return nil;

	_document = document;
	_resourceManager = [GLLResourceManager sharedResourceManager];
	
	itemDrawers = [[NSMutableArray alloc] init];
	
	NSEntityDescription *itemEntity = [NSEntityDescription entityForName:@"GLLItem" inManagedObjectContext:self.managedObjectContext];
	
	// Set up loading of future items and destroying items. Also update view.
	// Store self as weak in the block, so it does not retain this.
	__block __weak id weakSelf = self;
	managedObjectContextObserver = [[NSNotificationCenter defaultCenter] addObserverForName:NSManagedObjectContextObjectsDidChangeNotification object:self.managedObjectContext queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification *notification) {
		GLLSceneDrawer *self = weakSelf;
				
		NSMutableArray *toRemove = [[NSMutableArray alloc] init];
		for (GLLItemDrawer *drawer in self->itemDrawers)
		{
			if (![notification.userInfo[NSDeletedObjectsKey] containsObject:drawer.item])
				continue;
			
			[toRemove addObject:drawer];
		}
		[self->itemDrawers removeObjectsInArray:toRemove];
				
		// New objects includes absolutely anything. Restrict this to items.
		for (NSManagedObject *newItem in notification.userInfo[NSInsertedObjectsKey])
		{
			if ([notification.userInfo[NSDeletedObjectsKey] containsObject:newItem])
				continue; // Objects that were deleted again before this was called.
			if ([newItem.entity isKindOfEntity:itemEntity])
				[self _addDrawerForItem:(GLLItem *) newItem];
		}
		
		[self notifyRedraw];
	}];
    
    drawStateNotificationObserver = [[NSNotificationCenter defaultCenter] addObserverForName:GLLDrawStateChangedNotification object:nil queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification *notification) {
        [self notifyRedraw];
    }];
	
	// Load existing items
	NSFetchRequest *allItemsRequest = [[NSFetchRequest alloc] init];
	allItemsRequest.entity = itemEntity;
	
	NSArray *allItems = [self.managedObjectContext executeFetchRequest:allItemsRequest error:NULL];
	for (GLLItem *item in allItems)
		[self _addDrawerForItem:item];
	
	skeletonDrawer = [[GLLSkeletonDrawer alloc] initWithResourceManager:self.resourceManager];
	
	return self;
}

- (void)dealloc
{
	[[NSNotificationCenter defaultCenter] removeObserver:managedObjectContextObserver];
}

- (void)drawShowingSelection:(BOOL)showSelection into:(id<MTLRenderCommandEncoder>)commandEncoder  lightsBuffer:(id<MTLBuffer>)lights transformBuffer:(id<MTLBuffer>)transform
{
    GLLBeginTiming("Draw/Solid");
	// 1st pass: Draw items that do not need blending. They use shaders without alpha test
	for (GLLItemDrawer *drawer in itemDrawers)
        [drawer drawSolidInto:commandEncoder];
    GLLEndTiming("Draw/Solid");
    GLLBeginTiming("Draw/AlphaOp");
	
	/*// 2nd pass: Draw blended items, but only those pixels that are "almost opaque"
	glBindBufferBase(GL_UNIFORM_BUFFER, GLLUniformBlockBindingAlphaTest, self.resourceManager.alphaTestPassGreaterBuffer);
	
	glEnable(GL_BLEND);
	
	for (GLLItemDrawer *drawer in itemDrawers)
		[drawer drawAlphaInto:commandEncoder];
    
    GLLEndTiming("Draw/AlphaOp");
    GLLBeginTiming("Draw/AlphaTrans");
	// 3rd pass: Draw blended items, now only those things that are "mostly transparent".
	glBindBufferBase(GL_UNIFORM_BUFFER, GLLUniformBlockBindingAlphaTest, self.resourceManager.alphaTestPassLessBuffer);
	
	glDepthMask(GL_FALSE);
	for (GLLItemDrawer *drawer in itemDrawers)
        [drawer drawAlphaInto:commandEncoder];
    GLLEndTiming("Draw/AlphaTrans");*/
	
	if (showSelection)
    {
        GLLBeginTiming("Draw/Skel");
        
        [skeletonDrawer drawInto:commandEncoder];
        
        GLLEndTiming("Draw/Skel");
	}
}

- (NSManagedObjectContext *)managedObjectContext {
    return self.document.managedObjectContext;
}

#pragma mark - Selection

- (void)setSelectedBones:(NSArray<GLLItemBone *> *)selectedBones;
{
	skeletonDrawer.selectedBones = selectedBones;
	[self notifyRedraw];
}
- (NSArray<GLLItemBone *> *)selectedBones
{
	return skeletonDrawer.selectedBones;
}

#pragma mark - Private methods

- (void)_addDrawerForItem:(GLLItem *)item;
{
	NSError *error = nil;
    GLLItemDrawer *drawer = [[GLLItemDrawer alloc] initWithItem:item sceneDrawer:self error:&error];
	
	if (!drawer)
	{
		[NSApp presentError:error];
		if (item.objectID.isTemporaryID)
		{
			// Temporary ID means this was not loaded from a file. Get rid of it.
			[item.managedObjectContext deleteObject:item];
		}
		
		return;
    }
    
    if (drawer.replacedTextures.count > 0) {
        [self.document notifyTexturesNotLoaded:drawer.replacedTextures];
    }
	
	[itemDrawers addObject:drawer];
}

- (void)notifyRedraw;
{
	[[NSNotificationCenter defaultCenter] postNotificationName:GLLSceneDrawerNeedsUpdateNotification object:self];
}

@end

File: GLLara/GLLSceneDrawer.swift
//
//  GLLSceneDrawer.swift
//  GLLara
//
//  Created by Torsten Kammer on 06.07.22.
//  Copyright © 2022 Torsten Kammer. All rights reserved.
//

import Cocoa
import CoreData

@objc class GLLSceneDrawer: NSObject {
    
    @objc init(document: GLLDocument) {
        self.document = document
        skeletonDrawer = GLLSkeletonDrawer(resourceManager: GLLResourceManager.shared)
        
        super.init()
        
        managedObjectContextObserver = NotificationCenter.default.addObserver(forName: Notification.Name.NSManagedObjectContextObjectsDidChange, object: document.managedObjectContext, queue: OperationQueue.main) { [weak self] notification in
            guard let self = self else {
                return
            }
            
            let deletedObjects = notification.userInfo?[NSDeletedObjectsKey] as? NSSet
            if let deletedObjects = deletedObjects {
                self.itemDrawers.removeAll { drawer in
                    deletedObjects.contains(drawer.item)
                }
            }
            
            if let insertedObjects = notification.userInfo?[NSInsertedObjectsKey] as? NSSet {
                for newItem in insertedObjects {
                    if let deletedObjects = deletedObjects, deletedObjects.contains(newItem) {
                        continue;
                    }
                    
                    if let gllitem = newItem as? GLLItem {
                        self.addDrawer(for: gllitem);
                    }
                }
            }
            self.notifyRedraw()
        }
        
        drawStateNotificationObserver = NotificationCenter.default.addObserver(forName: Notification.Name.GLLDrawStateChanged, object: nil, queue: OperationQueue.main) { [weak self] notification in
            self?.notifyRedraw()
        }
        
        // Load existing items
        let allItemsRequest = NSFetchRequest<GLLItem>()
        allItemsRequest.entity = NSEntityDescription.entity(forEntityName: "GLLItem", in: document.managedObjectContext!)
        let allItems = try! document.managedObjectContext!.fetch(allItemsRequest)
        for item in allItems {
            addDrawer(for: item)
        }
        
    }
    
    deinit {
        if let observer = managedObjectContextObserver {
            NotificationCenter.default.removeObserver(observer)
        }
        if let observer = drawStateNotificationObserver {
            NotificationCenter.default.removeObserver(observer)
        }
    }
    
    weak var document: GLLDocument?
    
    @objc var managedObjectContext: NSManagedObjectContext? {
        return document?.managedObjectContext
    }
    @objc var resourceManager: GLLResourceManager {
        return GLLResourceManager.shared
    }
    
    @objc var selectedBones: [GLLItemBone] {
        set {
            skeletonDrawer.selectedBones = newValue
            notifyRedraw()
        }
        get {
            return skeletonDrawer.selectedBones
        }
    }
    
    func draw(into commandEncoder: MTLRenderCommandEncoder, blended: Bool) {
        for itemDrawer in itemDrawers {
            itemDrawer.draw(into: commandEncoder, blended: blended)
        }
    }
    
    func drawSelection(int commandEncoder: MTLRenderCommandEncoder) {
        skeletonDrawer.draw(into: commandEncoder)
    }
    
    func notifyRedraw() {
        NotificationCenter.default.post(name: NSNotification.Name.GLLSceneDrawerNeedsUpdate, object: self)
    }
    
    private var itemDrawers: [GLLItemDrawer] = []
    private let skeletonDrawer: GLLSkeletonDrawer
    private var drawStateNotificationObserver: Any? = nil
    private var managedObjectContextObserver: Any? = nil
    
    private func addDrawer(for item: GLLItem) {
        do {
            let drawer = try GLLItemDrawer(item: item, sceneDrawer: self)
            if drawer.replacedTextures.count > 0 {
                document!.notifyTexturesNotLoaded(drawer.replacedTextures)
            }
            itemDrawers.append(drawer)
        } catch {
            NSApp.presentError(error)
            if item.objectID.isTemporaryID {
                // Temporary ID means this was not loaded from a file. Get rid of it.
                item.managedObjectContext?.delete(item)
            }
        }
    }
    
}
